qiita_db/archive.py

```diff
@@ -0,0 +1,178 @@
+
+from __future__ import division
+
+import qiita_db as qdb
+
+
+class Archive(qdb.base.QiitaObject):
+    r"""Extra information for any features stored in a BIOM Artifact
+
+    Methods
+    -------
+    insert_from_biom
+
+    See Also
+    --------
+    qiita_db.QiitaObject
+    """
+
+    @classmethod
+    def _inserting_main_steps(cls, TRN, ms, features):
+        sql = """INSERT INTO qiita.archive_merging_scheme
+                    (archive_merging_scheme)
+                 SELECT %s WHERE NOT EXISTS (
+                    SELECT 1 FROM qiita.archive_merging_scheme
+                    WHERE archive_merging_scheme = %s)"""
+        TRN.add(sql, [ms, ms])
+        sql = """SELECT archive_merging_scheme_id
+                 FROM qiita.archive_merging_scheme
+                 WHERE archive_merging_scheme = %s"""
+        TRN.add(sql, [ms])
+        amsi = TRN.execute_fetchlast()
+
+        vals = [[amsi, _id, val] for _id, val in features.items()]
+        TRN.add("SELECT archive_upsert(%s, %s, %s)", vals, many=True)
+        TRN.execute()
+
+    @classmethod
+    def insert_from_artifact(cls, artifact, features):
+        r"""Inserts new features to the database based on a given artifact
+
+        Parameters
+        ----------
+        artifact : qiita_db.artifact.Artifact
+            The artifact from which the features were generated
+        features : dict {str: str}
+            A dictionary of the features and the values to be stored
+
+        Raises
+        ------
+            ValueError
+                If the Artifact type is not BIOM
+                If the artifact doesn't have a biom filepath
+        """
+        with qdb.sql_connection.TRN:
+            atype = artifact.artifact_type
+            if atype != 'BIOM':
+                raise ValueError(
+                    "To archive artifact must be BIOM but %s" % atype)
+
+            bfps = [fp for _, fp, fpt in artifact.filepaths if fpt == 'biom']
+            if not bfps:
+                raise ValueError("The artifact has no biom files")
+
+            # [0] as it returns a list
+            ms = qdb.util.get_artifacts_information(
+                [artifact.id])[0]['algorithm']
+
+            cls._inserting_main_steps(qdb.sql_connection.TRN, ms, features)
+
+    @classmethod
+    def get_merging_scheme_from_job(cls, job):
+        r"""Inserts new features to the database based on a given job
+
+        Parameters
+        ----------
+        job : qiita_db.artifact.Artifact
+            The artifact from which the features were generated
+        features : dict {str: str}
+            A dictionary of the features and the values to be stored
+
+        Raises
+        ------
+            ValueError
+                If the Artifact type is not BIOM
+                If the artifact doesn't have a biom filepath
+        """
+        with qdb.sql_connection.TRN:
+            acmd = job.command
```

Comment:
same?

Reply:
same ... 😄

Reply:
off topic but how does a developer know whether a given qiita method (or property) will issue a db call?

Reply:
I think the safest assumption is that any property/method from a Qiita object will issue a DB call ... as everything is store in there.